File: src/hw6/src/oscope.cpp
/*Example of publishing in ROS 2 for ROB499 HW2

oscope.cpp

Cory Hungate

This has been modified from sample code provided for the class by Bill Smart. 
I have kept some of the original comments that I find helpful as well as 
added my own
*/


#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/float32.hpp>

#include <chrono>
#include <cmath>

// constexpr double pi = 3.14159265358979323846;

using namespace std::chrono_literals;


// Create a class that inherits from Node.
class OscopePublisher : public rclcpp::Node {
public:
	
	OscopePublisher() :Node("oscope_publisher"), count_(0) {
		
		publisher_ = this->create_publisher<std_msgs::msg::Float32>("oscope", 10);

		timer_ = this->create_wall_timer(std::chrono::duration<double>(1.0 / sampling_rate_), std::bind(&OscopePublisher::timer_callback, this));
	}

private:
	// A variable to hold the current count.  The C++ idiom is to make private class variables
	// end in an underscore.
	double count_ = 0;
	double frequency_ = 1;
	double sampling_rate_ = 100;
	
	// Variables for the publisher and timer.  These shoudl be shared pointers.
	rclcpp::Publisher<std_msgs::msg::Float32>::SharedPtr publisher_;
	rclcpp::TimerBase::SharedPtr timer_;

	// The callback that the timer uses.
	void timer_callback() {
		// Make a new message, and set the data element. We can use the auto
		// keyword here, since the type is clear.  Note that we're assigning
		// the data element and post-incrementing the count_ variable.
		auto message = std_msgs::msg::Float32();
		message.data = sin(this->frequency_ * 2 * M_PI * this->count_);

		// Publish the message.
		publisher_->publish(message);

		this->count_ += 1/this->sampling_rate_;

		// Record in the log that we published the message. turning this off but leaving it in
		// RCLCPP_INFO(this->get_logger(), "Published %f %f", message.data, this->count_);
	}
};


// This is the entry point of the executable.
int main(int argc, char **argv) {
	// Initialize ROS.
	rclcpp::init(argc, argv);

	// Create a node and assign a shared pointer to it to the publisher variable.
	auto publisher = std::make_shared<OscopePublisher>();

	// Give control to the ROS event handler.
	rclcpp::spin(publisher);

	// Once the event handler is done, shut things down nicely.
	rclcpp::shutdown();

	return 0;
}

File: src/hw2/hw2/oscope.py
# Example of publishing in ROS 2 for ROB499 HW2
#
# oscope.py
#
# Cory Hungate
#
# This has been modified from sample code provided for the class by Bill Smart. 
# I have kept some of the original comments that I find helpful as well as 
# added my own

#the standard ROS imports + math
import rclpy
from rclpy.node import Node
import math

# my values are going to be float32 for this assignment
from std_msgs.msg import Float32

class BasicPublisher(Node):
	def __init__(self, wave_frequency):
		# Initializing the parent class and creating a publisher
		super().__init__('publisher')

		self.pub = self.create_publisher(Float32, 'oscope', 10)

		#creating a variable for sampling frequency and another for wave frequency
		self.sampling_frequency = 100
		self.wave_frequency = 2*math.pi*wave_frequency

		# I'm using the counter as my x values
		self.counter = 0.0

		#i'm using the timer as my x values, and going 
		self.timer = self.create_timer(1/self.sampling_frequency, self.callback)

		

	# This callback will be called every time the timer fires.
	def callback(self):
		# Make an Float32 message, calculate the information, and publish it
		msg = Float32()

		#sine wave = A*sin(omega*t + phi), for the base case of omega = 1 hz = 2pi rads
		msg.data = math.sin(self.wave_frequency * self.counter)
		self.pub.publish(msg)

		# Log that we published something.  In ROS2, loggers are associated with nodes, and
		# the idiom is to use the get_logger() call to get the logger.  This has functions
		# for each of the logging levels.
		self.get_logger().info(f'time = {self.counter} and y = {msg.data}')
		self.counter += (1/self.sampling_frequency)


# Creating a different function for each different sine wave I want to generate
# Each wave is basically just a copy/paste of the first with only the frequency changing
def basic_wave(args=None):
	# Initialize rclpy.  We should do this every time.
	rclpy.init(args=args)

	# Make a node class
	publisher = BasicPublisher(wave_frequency=1)

	# handing control over to ROS
	rclpy.spin(publisher)

	# Make sure we shutdown everything cleanly... just in case
	rclpy.shutdown()

#wave where the frequency is doubled (f=2)
def short_wave(args=None):

	rclpy.init(args=args)

	publisher = BasicPublisher(wave_frequency=2)

	rclpy.spin(publisher)

	rclpy.shutdown()

#wave where the frequency is halved (f=.5)
def long_wave(args=None):
	# Initialize rclpy.  We should do this every time.
	rclpy.init(args=args)

	# Make a node class
	publisher = BasicPublisher(wave_frequency=.5)

	# handing control over to ROS
	rclpy.spin(publisher)

	# Make sure we shutdown everything cleanly... just in case
	rclpy.shutdown()


# If we run the node as a script, then we're going to start here.
if __name__ == '__main__':
	# The idiom in ROS2 is to set up a main() function and to call it from the entry
	# point of the script.
	basic_wave()
